lib/puttyhosts/puttyhosts.go

```diff
@@ -0,0 +1,240 @@
+//     with a new hostname entry (from tsh puttyconfig <hostname>) which should be added to the list.
+//   - It appends the new hostname to the end of the hostList
+//   - All hostnames in the hostList are converted to their wildcard form if they contain a dot (test.example.com -> *.example.com)
+//     and are grouped together.
+//   - If a wildcard group only contains a single hostname which would be matched by its wildcard equivalent, that hostname is added
+//     to the hostList verbatim to prevent inadvertently matching against too many hosts with the same wildcard.
+//   - If a wildcard matches more than one hostname, the wildcard will be added to the hostList instead and the single hostnames
+//     discarded.
+//   - The hostList is then sorted alphabetically and returned.
+//
+// This is an effort to keep the length of hostList as short as possible for efficiency and tidiness, while not using any more
+// wildcards than necessary and preventing the need for end users to manually configure their trusted host CAs.
+func AddHostToHostList(hostList []string, hostname string) []string {
+	// add the incoming hostname to the hostList before we sort and process it
+	hostList = append(hostList, hostname)
+
+	hostMap := make(map[string][]string)
+	var extraHosts []string
+	// iterate over the full hostList
+	// if the element is a wildcard, add it to the list of wildcards
+	// if the element is not a wildcard, convert it to a wildcard and add any hostnames it matches to a map
+	for _, element := range hostList {
+		// FQDN-based hosts are grouped under a wildcard key
+		if hostnameContainsDot(element) {
+			wildcard := wildcardFromHostname(element)
+			if !slices.Contains(hostMap[wildcard], element) {
+				hostMap[wildcard] = append(hostMap[wildcard], element)
+			}
+		} else {
+			// any non-wildcard hosts go into the extraHosts list and will be processed separately
+			extraHosts = append(extraHosts, element)
+		}
+	}
+
+	var outputHostList []string
+	// first, add all non-wildcard matches separately
+	for _, hostname := range extraHosts {
+		if !slices.Contains(outputHostList, hostname) {
+			outputHostList = append(outputHostList, hostname)
+		}
+	}
+	// iterate over the map, look for all wildcard keys with more than one hostname matching.
+	// for each match, add the wildcard to the hostList.
+	for key, matchingHostnames := range hostMap {
+		// add all wildcards with more than one hostname matching
+		if len(matchingHostnames) > 1 {
+			outputHostList = append(outputHostList, key)
+		} else {
+			// add the single hostname which is matched by the given wildcard
+			outputHostList = append(outputHostList, matchingHostnames[0])
+		}
+	}
+
+	slices.Sort(outputHostList)
+	return outputHostList
+}
+
+var hostnameRegexp = regexp.MustCompile("^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]).)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9-]*[A-Za-z0-9])$")
```

Comment:
I was just curious, but the answer was quite educating: https://chat.openai.com/share/0898d082-ba3a-489e-a71c-83e84cbd3657

Reply:
This was actually really interesting; I'd never thought of using ChatGPT to explain or validate regex. It also gave me ideas for a few more test cases.

Unfortunately...
```
panic: regexp: Compile(`^(?:(?:(?:[a-zA-Z0-9](?:[a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?)\.){1,126}(?:[a-zA-Z0-9](?:[a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?))$`): error parsing regexp: invalid repeat count: `{1,126}`
```

Maybe being naive is a good thing?!